### PR TITLE
ERC20-Peg FrameV2 Upgrade

### PR DESCRIPTION
**Pull request by JasonTulp**
- Updates ERC20-Peg pallet to Frame V2
- Note, due to the introduction of BoundedVecs for Storage items ReadyBlocks and DelayedPaymentSchedule, there is a small possibility that some delayed payments may not be processed. Currently an event is thrown and an issue created [here](https://futureverse.atlassian.net/browse/TRN-397?atlOrigin=eyJpIjoiZTA5YTNjOTc3MzEzNGE4ZDhkNWZlZDE0YzYzOGNhYzIiLCJwIjoiaiJ9)

### CODE REVIEW

File: pallet/erc20-peg/src/lib.rs [hunk 1]
```diff
@@ -49,106 +48,212 @@ mod mock;
...
+			+ fungibles::Transfer<Self::AccountId, AssetId = AssetId, Balance = Balance>
+			+ fungibles::Mutate<Self::AccountId>;
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Interface to generate weights
+		type WeightInfo: WeightInfo;
+
+		/// The native token asset Id (managed by pallet-balances)
+		type NativeAssetId: Get<AssetId>;
```
Comment (surangap) [inline pallet/erc20-peg/src/lib.rs:103]:
```suggestion
                #[pallet::constant]
		type NativeAssetId: Get<AssetId>;
```

File: pallet/erc20-peg/src/lib.rs [hunk 2]
```diff
@@ -49,106 +48,212 @@ mod mock;
...
+		type MaxDelaysPerBlock: Get<u32>;
+
+		/// The maximum number of ready blocks allowed per storage
+		#[pallet::constant]
+		type MaxReadyBlocks: Get<u32>;
+	}
+
+	/// Whether deposit are active
+	#[pallet::storage]
+	#[pallet::getter(fn deposits_active)]
```
Comment (surangap) [inline pallet/erc20-peg/src/lib.rs:120]:
```suggestion
```
let's remove the getters since we are trying to move away from that

Comment (JasonTulp) [reply pallet/erc20-peg/src/lib.rs:120]: Good call, was planning to do this but got distracted

Comment (JasonTulp) [reply pallet/erc20-peg/src/lib.rs]: https://github.com/futureversecom/trn-seed/pull/825/commits/a1bc790e3764c6632fdf24cdb61ee1bdb4dd60aa

File: pallet/erc20-peg/src/lib.rs (hunk 1 above)
Comment (JasonTulp) [reply pallet/erc20-peg/src/lib.rs:102]: https://github.com/futureversecom/trn-seed/pull/825/commits/a1bc790e3764c6632fdf24cdb61ee1bdb4dd60aa

File: pallet/erc20-peg/src/lib.rs [hunk 3]
```diff
@@ -49,106 +48,194 @@ mod mock;
...
+
+	/// The ROOT peg contract address on Ethereum
+	#[pallet::storage]
+	pub type RootPegContractAddress<T> = StorageValue<_, EthAddress, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An erc20 deposit has been delayed.
+		Erc20DepositDelayed {
```
Comment (surangap) [inline pallet/erc20-peg/src/lib.rs:190]: nice... do you think downstreams catching these events will have work?

**Review by surangap — COMMENTED** [review]
lgtm

**Review by surangap — APPROVED** [review]
Have a benchmark run as well. let's keep it unmerged till ethy is also ready.

File: pallet/erc20-peg/src/lib.rs [hunk 4]
```diff
@@ -223,98 +303,155 @@ decl_module! {
...
+	impl<T: Config> Pallet<T> {
 		/// Activate/deactivate deposits (root only)
-		#[weight = T::WeightInfo::activate_deposits()]
-		pub fn activate_deposits(origin, activate: bool) {
+		#[pallet::weight(T::WeightInfo::activate_deposits())]
+		pub fn activate_deposits(origin: OriginFor<T>, activate: bool) -> DispatchResult {
 			ensure_root(origin)?;
-			DepositsActive::put(activate);
+			DepositsActive::<T>::put(activate);
+			Ok(())
```
Comment (zees-dev) [inline pallet/erc20-peg/src/lib.rs:336]: nit: oppurtunity to emit event here - upon state change

File: pallet/erc20-peg/src/lib.rs [hunk 5]
```diff
@@ -223,98 +303,155 @@ decl_module! {
...
 		}
 
 		/// Activate/deactivate withdrawals (root only)
-		#[weight = T::WeightInfo::activate_withdrawals()]
-		pub fn activate_withdrawals(origin, activate: bool) {
+		#[pallet::weight(T::WeightInfo::activate_withdrawals())]
+		pub fn activate_withdrawals(origin: OriginFor<T>, activate: bool) -> DispatchResult {
```
Comment (zees-dev) [inline pallet/erc20-peg/src/lib.rs:341]: nit: oppurtunity to emit event here - upon state change

File: pallet/erc20-peg/src/lib.rs (hunk 4 above)
Comment (zees-dev) [reply pallet/erc20-peg/src/lib.rs:336]: should have block number, and activate status

**Review by zees-dev — APPROVED** [review]
LGTM 👍

**Pull request merged**